File: java/apps/stats/src/Stats.java
public class Stats {
    /**
     * Returns the minimum value in the specified array.
     */
    public static double Min(double[] a) {
        double min = Double.POSITIVE_INFINITY;

        for (int i = 0; i < a.length; i++) {
            if (Double.isNaN(a[i])) {
                return Double.NaN;
	    }

            if (a[i] < min) {
                min = a[i];
	    }
	}

	return min;
    }

    /**
     * Returns the maximum value in the specified array.
     */
    public static double Max(double[] a) {
        double max = Double.NEGATIVE_INFINITY;

        for (int i = 0; i < a.length; i++) {
            if (Double.isNaN(a[i])) {
                return Double.NaN;
	    }

            if (a[i] > max) {
                max = a[i];
	    }
	}

	return max;
    }

    public static double Sum(double[] a) {
        double sum = 0.0;

        for (int i = 0; i < a.length; i++) {
            sum += a[i];
	}

        return sum;
    }

    public static double Mean(double[] a) {
        if (a.length == 0) {
            return Double.NaN;
	}

        double sum = Sum(a);

        return sum / a.length;  
    }

    /**
     * Returns the sample variance in the specified array.
     */
    public static double Variance(double[] a) {
        if (a.length == 0) {
            return Double.NaN;
	}

        double avg = Mean(a);
        double sum = 0.0;

        for (int i = 0; i < a.length; i++) {
            sum += (a[i] - avg) * (a[i] - avg);
	}

        return sum / (a.length - 1);
    }

    /**
     * Returns the sample standard deviation in the specified array.
     */
    public static double StandardDeviation(double[] a) {
        return Math.sqrt(Variance(a));
    }

    public static void main(String[] args) {
        double[] a = { 2.4, 2.3, 2.7, 2.6 };

	System.out.printf("Min: %10.2f\n", Min(a));
	System.out.printf("Max: %10.2f\n", Max(a));
	System.out.printf("Mean: %10.2f\n", Mean(a));
	System.out.printf("Variance: %10.2f\n", Variance(a));
	System.out.printf("Standard deviation: %10.2f\n", StandardDeviation(a));
    }
}

File: c/apps/moneybin/stats.c
#include "stats.h"

#include <math.h>

double stats_min(double const *data, size_t size)
{
  double min;
  size_t i;

  min = data[0];

  for (i = 1; i < size; i++) {
    if (data[i] < min) {
      min = data[i];
    }
  }

  return min;
}

double stats_max(double const *data, size_t size)
{
  double max;
  size_t i;

  max = data[0];

  for (i = 1; i < size; i++) {
    if (data[i] > max) {
      max = data[i];
    }
  }

  return max;
}

double stats_sum(double const *data, size_t size)
{
  double sum = 0;
  size_t i;

  for (i = 0; i < size; i++) {
    sum += data[i];
  }

  return sum;
}

double stats_avg(double const *data, size_t size)
{
  double sum = 0;

  sum = stats_sum(data, size);

  return sum / size;
}

double stats_var(double const *data, size_t size)
{
  double avg = 0;
  double sum = 0;
  size_t i;

  avg = stats_avg(data, size);

  for (i = 0; i < size; i++) {
    sum += (data[i] - avg) * (data[i] - avg);
  }

  return sum / (size - 1);
}

double stats_stdev(double const *data, size_t size)
{
  double var = 0;

  var = stats_var(data, size);

  return sqrt(var);
}
